### PR TITLE
Multi-Head Attention: Only require attn_mask if actually needed

### PR DESCRIPTION
**Pull request by f0k**
`torch.nn.MultiheadAttention` internally uses `scaled_dot_product_attention()`. For the common case of causal attention, the latter accepts an `is_causal` flag, which computes causal attention inside the kernel without having to compute an attention mask in memory. Still, `torch.nn.MultiheadAttention` and `torch.nn.functional.multi_head_attention_forward()` ask for an attention mask whenever `is_causal=True` is given:
```python
import torch
mha = torch.nn.MultiheadAttention(100, 4)
x = torch.random(2, 10, 100)
mha(x, x, x, is_causal=True, need_weights=False)  # RuntimeError
```
This short PR tightens the check for a missing attention mask so it is not required when it would be set to `None` 11 lines later anyway.
Disclaimer: I currently do not have a development setup for PyTorch and will rely on the CI, sorry.

As an aside, the docstring of `torch.nn.functional.multi_head_attention_forward()` currently reads:
> is_causal: If specified, applies a causal mask as attention mask, and ignores
>        attn_mask for computing scaled dot product attention.

This suggests that the mask is completely ignored, while it is actually still required when either `need_weights` or `key_padding_mask` is given. This could be fixed either by updating the docstring, or by creating a causal `attn_mask` on the fly when needed, and not ever complaining about a missing mask. The latter would be convenient, but it would hide to the user the opportunity to precompute the mask once and reuse it in the case of fixed sequence lengths or multiple same-size transformer layers.

### CODE REVIEW

**Comment by f0k** [comment]
Reading `MultiheadAttention.forward()`, I saw that I forgot the "fast path", which calls `torch._native_multi_head_attention` instead of `torch.nn.functional.multi_head_attention_forward()`. The fast path uses `attn_mask` without ever checking for `is_causal`. So calling `MultiheadAttention.forward(..., is_causal=True, attn_mask=None)` when all preliminaries are met for the fast path will currently not give an error, and not give the same result as when passing an attention mask.

The PR currently only touches `torch.nn.functional.multi_head_attention_forward()`, removing the need to pass an attention mask in exactly the cases it is currently guaranteed to be ignored anyway. By itself, this seems reasonable. However, it may lead users to pass `is_causal_True, attn_mask=None` to `MultiheadAttention.forward()`, which is a slippery slope -- it will work fine during training, but will break inference if the fast path is hit.

So in addition to updating the docstring for `torch.nn.functional.multi_head_attention_forward()`, I guess this PR should also clarify whether `MultiheadAttention.forward()` should require `attn_mask` when `is_causal=True`, and add the appropriate checks. My hunch would be to allow `is_causal=True, attn_mask=None`, but not enter the fast path in this case.
